posts/2021-10-05-configurable-response-headers-21.0.0.11-beta.adoc

```diff
@@ -0,0 +1,174 @@
+
+This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.
+
+If you try either package, <<feedback, let us know what you think>>.
+[#allbeta]
+== All Beta Features package
+
+The All Beta Features package includes the following beta features:
+
+* <<response_headers, Configurable Response Headers>>
+
+[#response_headers]
+=== Configurable Response Headers
+
+You can now configure Open Liberty (21.0.0.11-beta) to modify response headers. The available configuration options allow for headers to be appended, for existing headers to be overwritten, for missing headers to be added, and for undesired headers to be removed from all responses being serviced by an HTTP endpoint. This configuration offers more granular control over response headers, which offers a solution to modifying headers without the need to modify existing applications, filters, or otherwise. 
+
+==== Try it out! 
+
+To use configurable response headers, begin by defining a new element called `<headers>` in the `server.xml`.  You can configure this for individual HTTP endpoints or for all endpoints at once. 
+
+Configuring for individual HTTP endpoints:
+
+[source, xml]
+----
+<httpEndpoint id="defaultHttpEndpoint"
+                  httpPort="9080"
+                  httpsPort="9443" >
+	 
+       <headers>
+            <add>foo:bar</add>
+            <add>foo:bar2</add>
+            <set>customHeader:customValue</set>
+            <setIfMissing>X-Forwarded-Proto:https</setIfMissing>
+            <remove>Via</remove>
+       </headers>
+   		
+</httpEndpoint>
+----
+
+Configuring for all HTTP endpoints:
+
+[source, xml]
+----
+ <httpEndpoint id="defaultHttpEndpoint"
+                         httpPort="9080"
+                         httpsPort="9443"
+                         headersRef="myHeadersID">
+     </httpEndpoint>
+ 
+     <httpEndpoint id="otherHttpEndpoint"
+                         httpPort="9081"
+                         httpsPort="9444"
+                         headersRef="myHeadersID">
+    </httpEndpoint> 
+     
+     <headers id="myHeadersID">
+            <add>foo:bar</add>
+            <add>foo:bar2</add>
+            <set>customHeader:customValue</set>
+            <setIfMissing>X-Forwarded-Proto:https</setIfMissing>
+            <remove>Via</remove>
+     </headers>  
+----
+
+The `add` attribute allows multiple headers with the same name to be added to a response, similar to the HttpServletResponse's `addHeader` API. Similarly, the `set` attribute is analogous to the `setHeader` API, which sets a response header to the given name and value. This overwrites existing headers that share the same name. The `setIfMissing` attribute will only set the configured headers if they are not already present on the response. Lastly, the remove option will remove any response headers whose name matches a name defined by the configuration.
```

Comment:
suggested edit

The `setIfMissing` attribute will only set the configured headers if they are not already present on the response. Lastly, the remove option will remove any response headers whose name matches a name defined by the configuration.
--->
The `setIfMissing` attribute sets the configured headers only if they are not already present on the response. Lastly, the `remove` attribute removes any response headers with a name that matches the value that is specified in the attribute.